### ISSUE TITLE
CA1816: Call GC.SuppressFinalize correctly -- Redundant suggestion?

### ISSUE THREAD

**Issue by SetTrend** [dotnet/roslyn-analyzers#4745]
### Analyzer

**Diagnostic ID**: [CA1816](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1816)

### Describe the improvement

From the documentation:

> Violations of this rule can be caused by:
> - A method that is an implementation of [`IDisposable.Dispose`](https://docs.microsoft.com/dotnet/api/system.idisposable.dispose) and doesn't call [`GC.SuppressFinalize`](https://docs.microsoft.com/dotnet/api/system.gc.suppressfinalize).
> [...]
>
> To prevent derived types with finalizers from having to reimplement [`IDisposable`](https://docs.microsoft.com/dotnet/api/system.idisposable) and to call it, unsealed types without finalizers should still call [`GC.SuppressFinalize`](https://docs.microsoft.com/dotnet/api/system.gc.suppressfinalize).
<br/>

**I believe this part of the rule makes too many assumptions**
<br/>

### Describe suggestions on how to achieve the rule

The following cases may occur:

1. No class will derive from current class
2. A class with no finalizer will derive from current class
3. A class with finalizer but without disposing mechanism will derive from current class
3. A class with finalizer and with disposing mechanism will derive from current class

(a) Only case [4] will require a call to [`GC.SuppressFinalize`](https://docs.microsoft.com/dotnet/api/system.gc.suppressfinalize). In all other 3 cases, the rule will have no effect and add bare administrative burdon to the programmer.

(b) If the programmer of a derived class writes a finalizer and a disposing mechanism for its own unmanaged resources, (s)he will certainly be aware of the implications and want to add a call to [`GC.SuppressFinalize`](https://docs.microsoft.com/dotnet/api/system.gc.suppressfinalize) to be sure everything works fine. (S)he certainly will not rely on the author of a base class to have precautionary reconed with incidentally be derived by a class that's implementing a finalizer and a disposing mechanism and implemented [`GC.SuppressFinalize`](https://docs.microsoft.com/dotnet/api/system.gc.suppressfinalize) in the base class just in case.

**Comment by mavasani** [comment]
We need someone to investigate the current analyzer semantics and evaluate the proposed changes.

**Comment by sharwell** [comment]
> Only case [4] will require a call to GC.SuppressFinalize. In all other 3 cases, the rule will have no effect and add bare administrative burdon to the programmer.

The design guidelines for implementing `IDisposable` in public APIs (which CA1816 is part of enforcing) intentionally do not make any assumptions about the form of derived types. All derived types are considered the same in the fact that they might rely on the base class to provide the call to `GC.SuppressFinalize`.

> If the programmer of a derived class writes a finalizer and a disposing mechanism for its own unmanaged resources, (s)he will certainly be aware of the implications and want to add a call to GC.SuppressFinalize to be sure everything works fine.

This is not correct, because the implementation of `IDisposable.Dispose()` is not virtual. It is not possible for a derived type to correctly add the call to `GC.SuppressFinalize`, and it is also not necessary because the derived type may always _validly_ assume that the base type already provided the call.

**Comment by sharwell** [comment]
@SetTrend based on the information so far, the analyzer is behaving in accordance with the design guidelines it supports, and as such would not change behavior. If you have any other cases where you would propose a behavior change I would be happy to review those as well.

**Comment by SetTrend** [comment]
Thank you very much for taking the time to share this information with me, @sharwell!

I did some further investigation now.

Just to make sure, would you mind confirming that so, according to MS design guidelines, **the only valid way** of implementing a class hierarchy implementing `IDisposable` both in a base class and in a derived class is like this:

```c#
using System;
using System.Diagnostics;

namespace FinalizeTest
{
  public class BaseClass : IDisposable
  {
    public BaseClass() => Debug.Print("BaseClass constructor.");

    ~BaseClass()
    {
      Debug.Print("BaseClass finalizer.");

      Dispose(false);
    }

    public void Dispose()
    {
      Debug.Print("BaseClass Dispose.");

      Dispose(true);

      Debug.Print("BaseClass SuppressFinalize.");
      GC.SuppressFinalize(this);
    }

    public void Dispose(bool disposing) => Debug.Print($"BaseClass Disposing(${disposing}).");
  }
}
```
```c#
using System;
using System.Diagnostics;

namespace FinalizeTest
{
  public class DerivedClass : BaseClass, IDisposable
  {
    public DerivedClass() => Debug.Print("DerivedClass constructor.");

    ~DerivedClass()
    {
      Debug.Print("DerivedClass finalizer.");

      Dispose(false);
    }

    public new void Dispose()
    {
      Debug.Print("DerivedClass Dispose.");

      Dispose(true);

      Debug.Print("DerivedClass SuppressFinalize.");
      GC.SuppressFinalize(this);
    }

    public new void Dispose(bool disposing)
    {
      Debug.Print($"DerivedClass Disposing(${disposing}).");

      base.Dispose(disposing);
    }
  }
}
```
```c#
namespace FinalizeTest
{
  class Program
  {
    static void Main(string[] _)
    {
      using BaseClass b = new DerivedClass();
    }
  }
}
```

Having ...
1. finalizers optional in both classes
2. only `Dispose(bool disposing)` calling its base class counterpart
3. `GC.SuppressFinalize()` being mandatory to each `Dispose()` method in the hierarchy

**Comment by SetTrend** [comment]
Never mind. I found the explanation here:

https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#implement-the-dispose-pattern-for-a-derived-class

**Comment by sharwell** [comment]
@SetTrend No, there are several problems with the sample above.

1. Setting aside the general recommendation that you should _never_ write a finalizer, `DerivedClass` should not add a finalizer if `BaseClass` already adds one.
2. `DerivedClass` must inherit the implementation of `IDisposable.Dispose()` from `BaseClass`. In other words, `DerivedClass.Dispose()` needs to be deleted.
3. The implementation of `Dispose()` in `BaseClass` should not contain any code except for the call to `Dispose(true)` and the call to `SuppressFinalize()`.
4. `DerivedClass` may _override_ the implementation of `Dispose(bool)` it inherits from `BaseClass`, but it should not _hide_ it. In other words, the `new` keyword needs to be changed to `override`.
5. `Dispose(bool)` needs to be `protected`, not `public`.
6. Finalizers are not allowed to use managed resources from other types. For the case above, the behavior `Debug.Print` within the finalizers is undefined.

**Comment by SetTrend** [comment]
You've been faster than me. I already deleted my comments when I hit the MS docs article when your comment appeared.

I very much shy away from bothering your further, but given your explanations, what if a hypothetical derived class derived from, say `FileStream` (which implements `IDisposable`). If the derived class added unmanaged private resources to perform its duties (e.g. logging some information to a file), wouldn't it need to implement a finalizer in case the program won't call `Dispose()` on an object created from this class?

The following example I doodled for this question (I'll be using a managed `StreamWriter` private field for brevity here):

```c#
using System.Diagnostics;
using System.IO;

namespace FinalizeTest
{
  public class FileStreamLogger : FileStream
  {
    private StreamWriter _logFile;



    public FileStreamLogger(string path, FileMode mode) : base(path, mode)
    {
      Debug.Print("FileStreamLogger constructor.");

      _logFile = new StreamWriter(path + ".log");
    }

    ~FileStreamLogger()
    {
      Dispose(false);
    }

    public new void Dispose()
    {
      Debug.Print("FileStreamLogger Dispose.");

      Dispose(true);

      Debug.Print("FileStreamLogger SuppressFinalize.");
    }

    protected override void Dispose(bool disposing)
    {
      base.Dispose(disposing);

      _logFile?.Close();
      _logFile = null;
    }
  }
}
```

**Comment by SetTrend** [comment]
Ah .. Again I see your point only after I've been writing my question. The `Dispose(bool disposing)` is overridden, of course. So the base class finalizer will call my `Dispose(bool disposing)` method and release my unmanaged resources.

Please pardon me for bothering you :blush:

**Comment by sharwell** [comment]
@SetTrend it's no problem. This is a super complicated topic so questions tend to come up. One of the reasons why the rules are so restrictive on options is it helps users who just want to work on something else and aren't looking to understand the details. 😄 But for people that do want to learn the motivation behind the design we're happy to help out.
